pkg/controller/registry/reconciler.go

```diff
@@ -0,0 +1,360 @@
+	if err != nil {
+		logrus.WithField("roleBinding", roleBindingName).WithError(err).Warn("couldn't find role binding in cache")
+		return nil
+	}
+	return roleBinding
+}
+
+func (c *ConfigMapRegistryReconciler) currentPods(source catalogSourceDecorator, image string) []*v1.Pod {
+	podName := source.Pod(image).GetName()
+	pods, err := c.Lister.CoreV1().PodLister().Pods(source.GetNamespace()).List(source.Selector())
+	if err != nil {
+		logrus.WithField("pod", podName).WithError(err).Warn("couldn't find pod in cache")
+		return nil
+	}
+	if len(pods) > 1 {
+		logrus.WithField("selector", source.Selector()).Warn("multiple pods found for selector")
+	}
+	return pods
+}
+
+func (c *ConfigMapRegistryReconciler) currentPodsWithCorrectResourceVersion(source catalogSourceDecorator, image string) []*v1.Pod {
+	podName := source.Pod(image).GetName()
+	pods, err := c.Lister.CoreV1().PodLister().Pods(source.GetNamespace()).List(labels.SelectorFromValidatedSet(source.Labels()))
+	if err != nil {
+		logrus.WithField("pod", podName).WithError(err).Warn("couldn't find pod in cache")
+		return nil
+	}
+	if len(pods) > 1 {
+		logrus.WithField("selector", source.Selector()).Warn("multiple pods found for selector")
+	}
+	return pods
+}
+
+// Ensure that all components of registry server are up to date.
+func (c *ConfigMapRegistryReconciler) EnsureRegistryServer(catalogSource *v1alpha1.CatalogSource) error {
+	source := catalogSourceDecorator{catalogSource}
+
+	// fetch configmap first, exit early if we can't find it
+	configMap, err := c.Lister.CoreV1().ConfigMapLister().ConfigMaps(source.GetNamespace()).Get(source.Spec.ConfigMap)
+	if err != nil {
+		return fmt.Errorf("unable to get configmap %s/%s from cache", source.GetNamespace(), source.Spec.ConfigMap)
+	}
+
+	if source.ConfigMapChanges(configMap) {
+		catalogSource.Status.ConfigMapResource = &v1alpha1.ConfigMapResourceReference{
+			Name:            configMap.GetName(),
+			Namespace:       configMap.GetNamespace(),
+			UID:             configMap.GetUID(),
+			ResourceVersion: configMap.GetResourceVersion(),
+		}
+	}
+
+	// if service status is nil, we force create every object to ensure they're created the first time
+	overwrite := source.Status.RegistryServiceStatus == nil
+
+	// recreate the pod if there are configmap changes; this causes the db to be rebuilt
+	// recreate the pod if no existing pod is serving the latest configmap
+	overwritePod := overwrite || source.ConfigMapChanges(configMap) || len(c.currentPodsWithCorrectResourceVersion(source, c.Image)) == 0
+
+	//TODO: if any of these error out, we should write a status back (possibly set RegistryServiceStatus to nil so they get recreated)
+	if err := c.ensureServiceAccount(source, overwrite); err != nil {
+		return err
+	}
+	if err := c.ensureRole(source, overwrite); err != nil {
+		return err
+	}
+	if err := c.ensureRoleBinding(source, overwrite); err != nil {
+		return err
+	}
+	if err := c.ensurePod(source, overwritePod); err != nil {
```

Comment:
We could make use of a Deployment instead of a Pod. Then we would get rolling updates and a safeguard against losing a working registry pod (if the CatalogSource's ConfigMap is no longer valid).